toffy/normalize.py

```diff
@@ -0,0 +1,238 @@
+# adapted from https://machinelearningmastery.com/curve-fitting-with-python/
+
+import json
+import os
+import shutil
+import warnings
+
+import numpy as np
+from scipy.optimize import curve_fit
+import skimage.io as io
+import matplotlib.pyplot as plt
+import pandas as pd
+
+from ark.utils import io_utils, load_utils
+
+
+def create_objective_function(obj_func):
+    """Creates a function of specified type to be used for fitting a curve
+
+    Args:
+        obj_func (str): the desired objective function. Must be either poly_2, ..., poly_5, or log
+
+    Returns:
+        function: the function which will be optimized"""
+
+    valid_funcs = ['poly_2', 'poly_3', 'poly_4', 'poly_5', 'log', 'exp']
+    if obj_func not in valid_funcs:
+        raise ValueError('Invalid function, must be one of {}'.format(valid_funcs))
+
+    if obj_func == 'poly_2':
+        def objective(x, a, b, c):
+            return a * x + b * x ** 2 + c
+    elif obj_func == 'poly_3':
+        def objective(x, a, b, c, d):
+            return a * x + b * x ** 2 + c * x ** 3 + d
+    elif obj_func == 'poly_4':
+        def objective(x, a, b, c, d, e):
+            return a * x + b * x ** 2 + c * x ** 3 + d * x ** 4 + e
+    elif obj_func == 'poly_5':
+        def objective(x, a, b, c, d, e, f):
+            return a * x + b * x ** 2 + c * x ** 3 + d * x ** 4 + e * x ** 5 + f
+    elif obj_func == 'log':
+        def objective(x, a, b):
+            return a * np.log(x) + b
+    elif obj_func == 'exp':
+        def objective(x, a, b, c, d):
+            x_log = np.log(x)
+            return a * x_log + b * x_log ** 2 + c * x_log ** 3 + d
+
+    return objective
+
+
+def fit_calibration_curve(x_vals, y_vals, obj_func, plot_fit=False):
+    """Finds the optimal weights to fit the supplied values for the specified function
+
+    Args:
+        x_vals (list): the x values to be fit
+        y_vals (list): the y value to be fit
+        obj_func (str): the name of the function that will be fit to the data
+        plot_fit (bool): whether or not to plot the fit of the function vs the values
+
+    Returns:
+        list: the weights of the fitted function"""
+
+    # get objective function
+    objective = create_objective_function(obj_func)
+
+    # get fitted values
+    popt, pcov = curve_fit(objective, x_vals, y_vals)
```

Comment:
With a `dict` of `lambda` objective functions, you can use `popt, _ = curve_fit(obj_functions[obj_func], x_vals, y_vals)`. Unless you plan on using `pcov` in a later PR I think it can safely be dropped.